chunk/dynamo_table_manager.go

```diff
@@ -0,0 +1,347 @@
+)
+
+func init() {
+	prometheus.MustRegister(tableCapacity)
+}
+
+// TableManagerConfig is the config for a DynamoTableManager
+type TableManagerConfig struct {
+	DynamoDBURL          string
+	DynamoDBPollInterval time.Duration
+
+	UsePeriodicTables    bool
+	TablePrefix          string
+	TablePeriod          time.Duration
+	PeriodicTableStartAt time.Time
+
+	// duration a table will be created before it is needed.
+	CreationGracePeriod        time.Duration
+	MaxChunkAge                time.Duration
+	ProvisionedWriteThroughput int64
+	ProvisionedReadThroughput  int64
+
+	// Not exported as only used by tests to inject mocks
+	dynamodb dynamodbClient
+}
+
+// DynamoTableManager creates and manages the provisioned throughput on DynamoDB tables
+type DynamoTableManager struct {
+	cfg      TableManagerConfig
+	dynamodb dynamodbClient
+
+	// the name of the single, pre-weekly table
+	tableName string
+
+	done chan struct{}
+	wait sync.WaitGroup
+}
+
+// NewDynamoTableManager makes a new DynamoTableManager
+func NewDynamoTableManager(cfg TableManagerConfig) (*DynamoTableManager, error) {
+	dynamodbClient, tableName := cfg.dynamodb, ""
+	if dynamodbClient == nil {
+		dynamodbURL, err := url.Parse(cfg.DynamoDBURL)
+		if err != nil {
+			return nil, err
+		}
+
+		dynamoDBConfig, err := awsConfigFromURL(dynamodbURL)
+		if err != nil {
+			return nil, err
+		}
+
+		dynamodbClient = dynamoClientAdapter{dynamodb.New(session.New(dynamoDBConfig))}
+		tableName = strings.TrimPrefix(dynamodbURL.Path, "/")
+	}
+
+	m := &DynamoTableManager{
+		dynamodb:  dynamodbClient,
+		tableName: tableName,
+
+		cfg:  cfg,
+		done: make(chan struct{}),
+	}
+	return m, nil
+}
+
+// Start starts the DynamoTableManager
+func (m *DynamoTableManager) Start() {
+	m.wait.Add(1)
+	go m.loop()
+}
+
+// Stop stops the DynamoTableManager
+func (m *DynamoTableManager) Stop() {
+	close(m.done)
+	m.wait.Wait()
+}
+
+func (m *DynamoTableManager) loop() {
+	defer m.wait.Done()
+
+	ticker := time.NewTicker(m.cfg.DynamoDBPollInterval)
+	defer ticker.Stop()
+
+	if err := instrument.TimeRequestHistogram(context.Background(), "DynamoTableManager.syncTables", syncTableDuration, func(ctx context.Context) error {
+		return m.syncTables(ctx)
+	}); err != nil {
+		log.Errorf("Error syncing tables: %v", err)
+	}
+
+	for {
+		select {
+		case <-ticker.C:
+			if err := instrument.TimeRequestHistogram(context.Background(), "DynamoTableManager.syncTables", syncTableDuration, func(ctx context.Context) error {
+				return m.syncTables(ctx)
+			}); err != nil {
+				log.Errorf("Error syncing tables: %v", err)
+			}
+		case <-m.done:
+			return
+		}
+	}
+}
+
+func (m *DynamoTableManager) syncTables(ctx context.Context) error {
+	expected := m.calculateExpectedTables(ctx)
+
+	toCreate, toCheckThroughput, err := m.partitionTables(ctx, expected)
+	if err != nil {
+		return err
+	}
+
+	if err := m.createTables(ctx, toCreate); err != nil {
+		return err
+	}
+
+	return m.updateTables(ctx, toCheckThroughput)
```

Comment:
No reason, I think its easiest just to fail everything - we'll need monitoring / alerting on this either way.